Feature/document bsm and lists permission endpoints (#46)
documaster/noark5-web-services: version1/04 - Web services.md

**Response**

204 No Content

## Managing code list permissions

This set of endpoints allows you to manage the explicit permissions on the specified objects:
* all code lists
* just the administrativEnhet code list or any of its list values
* just the mappetype code list or any of its list values
* just the dokumenttype code list or any of its list values
* just the skjerming code list or any of its list values

Address:
```
https://{server}:{port}/rms/api/public/noark5/v1/permission/code-list
```

### Get existing permissions

Gets existing permissions for all code lists, a single code list, or a specific code list value.

The use cases are, respectively:

> * GET /rms/api/public/noark5/v1/permission/code-list?listType=all

Reviewer: Why not just `GET /rms/api/public/noark5/v1/permission/code-list` as in not specify the listType parameter?

Reviewer: I think I thought of something else while first seeing the endpoint, but now I understand what 'all' is. Still, sounds a bit confusing that 'all' is the parent of all code lists, but not the set of code lists. I currently don't know what else to suggest.

Author: The underlying issue here is the fact that we don't treat code lists as resources in v1.

I was facing exactly the same challenge when I started on this and I was thinking exactly the same thing you are suggesting. However, we already have a `GET /code-lists` endpoint which has nothing to do with the parent of all lists and if I made this `GET /code-list`, I think this would have simply added to the problem, not solved it.

That's my reasoning and that's why I went with the documentation you're seeing now. I don't have any better suggestions either, but I decided that if I explicitly document the purpose of "all", it'd be better than treating just `code-list` as the parent of all lists.

The only thing that I could do (in the context of the already developed web services) is to probably add a "/code-list-parent" endpoint or the likes, but that seemed superfluous since I could simply create a new type of code list (all).

- **code** (optional)
  - the specific code list value for which to fetch existing permissions
  - when present, only the permissions applied on the specified code list value will be returned
- **offset** (optional)
  - offset of first result
  - defaults to 0
  - must be greater than or equal to 0
- **limit** (optional)
  - maximum number of results to retrieve
  - must be greater than 0 and less than or equal to 200
  - defaults to 10

**Response**

200 OK

```
Content-type: application/json

{
    "permissions": [
      {
        "groupId": long,
        "listType": string,
        "code": string,
        "explicitPermissions": [string]   
      }
    ],
    "hasMore": boolean
}
```

###### Details

- **groupId**
  - the group ID for which permissions were assigned
- **listType**
  - the list type for which permissions were assigned
  - could be one of 'all', 'admnistrativEnhet', 'mappetype', 'dokumenttype', 'skjerming', where 'all' represents permissions applicable to all code lists and the other values represent permissions applied on a specific code list
  - 'all' can be considered as an entity of its own that is the parent of existing code lists
- **code**

>   - the specific code list value for which to fetch existing permissions

Reviewer: ```suggestion
  - the specific code list value for which permissions were assigned
```

  - could be one of 'all', 'admnistrativEnhet', 'mappetype', 'dokumenttype', 'skjerming', where 'all' represents permissions applicable to all code lists and the other values represent permissions applied on a specific code list
  - 'all' can be considered as an entity of its own that is the parent of existing code lists
- **code** (optional)
  - the specific code list value for which to fetch existing permissions
  - when present, only the permissions applied on the specified code list value will be returned
- **offset** (optional)
  - offset of first result
  - defaults to 0
  - must be greater than or equal to 0
- **limit** (optional)
  - maximum number of results to retrieve
  - must be greater than 0 and less than or equal to 200
  - defaults to 10

**Response**

200 OK

```
Content-type: application/json

{
    "permissions": [
      {
        "groupId": long,
        "listType": string,
        "code": string,
        "explicitPermissions": [string]   
      }
    ],
    "hasMore": boolean
}
```

###### Details

- **groupId**
  - the group ID for which permissions were assigned
- **listType**
  - the list type for which permissions were assigned

>   - could be one of 'all', 'admnistrativEnhet', 'mappetype', 'dokumenttype', 'skjerming', where 'all' represents permissions applicable to all code lists and the other values represent permissions applied on a specific code list

Reviewer 2: Can we really return _all_ in the response? If we can return _all_ then I guess the code field below will be omitted?

Author: That is correct. When "all" is used, the code field is never present.

  - 'all' can be considered as an entity of its own that is the parent of existing code lists
- **code** (optional)
  - the specific code list value for which to create permissions
- **explicitPermissions**
  - a non-empty list of explicit permissions to set
  - see "available service and explicit permissions" endpoint or the Access Control document

**Response**

201 Created

```
Content-type: application/json

{
    "groupId": long,
    "listType": string,
    "code": string,
    "explicitPermissions": [string]   
}
```

###### Details

- **groupId**
  - the group ID for which permissions were assigned
- **listType**
  - the list type for which permissions were assigned
  - could be one of 'all', 'admnistrativEnhet', 'mappetype', 'dokumenttype', 'skjerming', where 'all' represents permissions applicable to all code lists and the other values represent permissions applied on a specific code list
  - 'all' can be considered as an entity of its own that is the parent of existing code lists

> - **code**

Reviewer: ```suggestion
- **code** (optional)
```

>   - the specific code list value for which to fetch existing permissions

Reviewer: ```suggestion
  - the specific code list value for which permissions were assigned
```

  - 'all' can be considered as an entity of its own that is the parent of existing code lists
- **code** (optional)
  - the specific code list value whose permissions to update
- **explicitPermissions**
  - a non-empty list of explicit permissions to update
  - see "available service and explicit permissions" endpoint or the Access Control document

**Response**

200 OK

```
Content-type: application/json

{
    "groupId": long,
    "listType": string,
    "code": string,
    "explicitPermissions": [string]   
}
```

###### Details

- **groupId**
  - the group ID for which permissions were assigned
- **listType**
  - the list type for which permissions were assigned
  - could be one of 'all', 'admnistrativEnhet', 'mappetype', 'dokumenttype', 'skjerming', where 'all' represents permissions applicable to all code lists and the other values represent permissions applied on a specific code list
  - 'all' can be considered as an entity of its own that is the parent of existing code lists

> - **code**

Reviewer: ```suggestion
- **code** (optional)
```

>   - the specific code list value for which to fetch existing permissions

Reviewer: ```suggestion
  - the specific code list value for which permissions were assigned
```

Authorization: Bearer ACCESS_TOKEN
```

###### Details

- **groupId** (optional)
  - the group ID for which to delete permissions
  - if the groupId is omitted, all permissions on the specified object will be deleted
- **listType**
  - the list type whose permissions to delete
  - could be one of 'all', 'admnistrativEnhet', 'mappetype', 'dokumenttype', 'skjerming', where 'all' represents permissions applicable to all code lists and the other values represent permissions applied on a specific code list
  - 'all' can be considered as an entity of its own that is the parent of existing code lists
- **code** (optional)
  - the specific code list value whose permissions to delete

**Response**

204 No Content

## Managing BSM registry permissions

This set of endpoints allows you to manage the explicit permissions on the BSM registry.

Address:
```
https://{server}:{port}/rms/api/public/noark5/v1/permission/bsm-registry
```

### Get existing permissions

Gets existing permissions for the BSM registry.

**Request**

``` text
GET /rms/api/public/noark5/v1/permission/bsm-registry?groupId=LONG&offset=INTEGER&limit=INTEGER HTTP/1.1
Authorization: Bearer ACCESS_TOKEN
```

###### Details

> - **groupId** (optional)

Reviewer: This is the access group ID, right? Not the BSM group ID?
Could be misleading.

Author: Hmm, that's correct and you have a point.

Do you think it would make more sense to modify the name or the documentation would be enough?

Reviewer: I would say renaming the property would be best. But then would you rename it for the BSM registry endpoints only, or for the entity/code list, too?

Author: For consistency, it will have to be renamed in all endpoints. Otherwise, it would be even more confusing.

Reviewer: Yes, that would be best, at least in my opinion.

Author: We've agreed to rename it. I'll do it in a separate PR though.

- **hasMore**
  - indicates if there are more results matching the query than those in the current page

### Create a new permission

Creates a new permission for the BSM registry.

**Request**

``` text
POST /rms/api/public/noark5/v1/permission/bsm-registry HTTP/1.1
Authorization: Bearer ACCESS_TOKEN
Content-Type: application/json

{
    "groupId": long,
    "explicitPermissions": [string]   
}
```

###### Details

- **groupId**
  - the group ID for which to assign permissions
- **explicitPermissions**
  - a non-empty list of explicit permissions to set
  - see "available service and explicit permissions" endpoint or the Access Control document

**Response**

201 Created

```
Content-type: application/json

{
    "groupId": long,
    "explicitPermissions": [string]   
}
```

###### Details

- **groupId**
  - the group ID for which permissions were assigned
- **explicitPermissions**
  - a non-empty list of assigned explicit permissions
  - see "available service and explicit permissions" endpoint or the Access Control document

### Update an existing permission

Updates an existing permission for the BSM registry.

**Request**

``` text

> PUT /rms/api/public/noark5/v1/permission/bsm-registry HTTP/1.1

Reviewer: Will the PUT assume that some kind of permissions are already specified for this access group?

Author: It definitely will. We require that you can call "update" only for existing permissions. We observe the same behavior for entity and code list permissions.

What made you ask this question by the way and what would you expect as a change? (since I see that this is not different in any way from the documentation about entity and code list permissions)

Reviewer: It's not different, yes, and it is fine this way. I have not seen the actual implementation, yet, decided to start with the documentation first this time :D
